rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,263 @@
+to serve as the backing of the gRPC API, and make internal decisions as if it were a remote PDP. This PDP will be a shared
+service initialized once at process startup.
+
+On agents, a local ephemeral PDP will be able to be initialized with a fake user store only knowing the identity of the
+user making the request. This variant of the PDP will need to be initialized for each incoming request, and will be able to
+only serve requests related to the user for which it was initialized (much like the `AccessChecker` today).
+
+Common PDP logic will be shared between the two variants, meaning that so long as we need to support the agent-local PDP,
+the control-plane PDP won't be able to make use of any state that cannot be derived from a combination of local agent cache
+and user certificate identity. The APIs of both local and remote PDP will conform to the same interface so that any logic
+relying upon the PDP can abstract over local and remote implementations.
+
+During this phase we want to start moving as many elements as possible from being part of the certificate-derived identity to
+being calculated on-demand within the PDP itself. As a prerequisite to this, we will need to implement a custom access request
+cache, as access requests have a signficant impact on many identity fields. A custom cache will be necessary in order to handle
+the fact that access requests are often used immediately upon approval, before approval can be relied upon to be propagated
+to caches naturally.
+
+The intent of the above work is to allow us to make the vast majority of necessary code changes in a totally backwards-compatible
+manner and to backport them to all active release branches. This will ensure that ongoing work related to policy enforcement
+doesn't need to reimplement the same logic for the old model when backporting. Though its worth considering trying to land
+all major changes prior to a testplan so that we can get robust manual testing of all major features with the changes in
+place before backporting.
+
+To ensure that we've fully broken agent dependence on roles outside of the PDP, we will split the agent's local cache up
+s.t. all components other than the PDP will be statically prevented from accessing roles.
+
+During this phase we will also implement tctl commands for directly invoking the DecisionService gRCP API on the control
+plane. This will provide a powerful debugging and auditing tool for superusers and developers who want deeper insight into how
+teleport makes decisions.
+
+#### Relocate Phase
+
+The relocate phase will see us actually transition to all access-control decisions being made at the control plane. The method
+of relocation will depend on the kind of teleport agent. For tunnel agents, we will start performing denial at the control plane
+itself, and allowed dials will have the Permit object forwarded to the agent as part of the dial. For direct dial agents, the
```

Comment:
Could we use this change to also prohibit direct dial that bypasses the proxy? ~~Or maybe deprecate direct dial altogether? 🥺~~

The Teleport SSHD listener is already specialized to at least opportunistically understand additional data signed by the proxy as part of client IP propagation, if we don't have to support openssh connecting directly to a server's 3022 we'd avoid the extra roundtrip.

Reply:
> if we don't have to support openssh connecting directly to a server's 3022 we'd avoid the extra roundtrip.

I'd ask that we don't take this decision too lightly. On the Machine ID team, we fairly regularly get folks asking for the ability to direct connect to a reverse tunnel agent! This is typically for two reasons:

- Performance: Why make a trip out of the datacenter to where your Teleport Proxies are, when the two machines trying to connect to one another are in the same datacenter?
- Performance: Why swamp your Proxy with large amounts of traffic (e.g copying a large binary from an Ansible master to a fleet of machines) and potentially impact other Teleport users?
- Reliability: Ability for M2M SSH to succeed even if Proxy is unavailable/network is unreliable.

For the most part, we've been able to push people back on this and make perf improvements to Teleport itself - but - I do think one day we may hit a limitation here. It'd be nice to keep "machines" in mind here if possible!

Reply:
The benefits of a PDP system are mostly negated if we still leave a mechanism in place for it to be bypassed (would mean agents still need to replicate and understand all access-control related resources, and we'd need to still restrict ourselves to only implementing features that are representable as a combination of cert state and roles).

If we don't preserve a mechanism for the PDP to be bypassed, then even if you direct dial an agent you still have to wait on a network round-trip to the control plane to get a decision from the PDP... so the reliability and perf points are mostly negated.  There would still be potential benefits to long-running and/or heavy workloads (e.g. file transfer, or anything super latency sensitive) of course.  IDK how strongly we weight those, but there might be alternative ways to address those cases that let us still adopt a more consistent PDP architecture.  Ex:

- Connection Upgrades: we could support a flow for upgrading a proxied connection into a direct connection under certain circumstances.  This would ensure the PDP is always inline, but can permit connections to become out of line as appropriate.
- Intrusive Proxies: this is already *kinda* possible in some cases, but we could improve support for running a local proxy inside of your datacenter that allows all datacenter-local operations to stay within the datacenter without the need to support multiple separate trust flows.

Reply:
> The benefits of a PDP system are mostly negated if we still leave a mechanism in place for it to be bypassed

Agreed - I am not advocating for bypassing PDP. I am advocating for it being possible for the agent to act as the PEP for direct dial connections.

> so the reliability and perf points are mostly negated. 

I concede that yes, this does mean that PEP must be available for the client to be able to directly connect to the server - which negates the reliability aspect. However, I don't think this negates performance benefits? Common management tools like Ansible are pretty much the opposite of light-weight and are extremely noisy. At least in my experience so far, latency between the master and the server can make a significant difference in the amount of time that a playbook takes to run across hundreds or thousands of hosts. Ansible playbooks also tend to be "fragile" to failure. So even if the latency is not a major concern, the success rate of a large/complex ansible playbook is impacted by the connection travelling out to the proxy and back in.

> IDK how strongly we weight those

I'm a little biased - I feel like we've been 'underweighing' these so far and that this has had negative impact on large deals and customer experience. If we don't consider "machine scale" more often in our decision making, then we're likely to continue introducing significant regressions which are unnoticeable at "human scale" but are show-stoppers for large customers.

This perhaps isn't an argument for a specific choice/direction here - but just a more general plea that we try to think about this more.

> alternative ways to address those cases that let us still adopt a more consistent PDP architecture

Definitely open to these ideas. I don't think I'm dying on the hill of "we must keep direct connections", just keen that we ensure that we're planning for this long-term.